### PR TITLE
`mixed16` mode and its support in `L.BatchNormalization`

### PR DESCRIPTION
**Pull request by beam2d**
This PR introduces a new default dtype mode named `mixed16`. With this mode, Chainer basically uses `float16` as the default float dtype, but when Chainer decides that it is better to use higher precision representation for numerical stability and computational efficiency, it uses other float dtypes, esp. `float32`.

The batch normalization fix is based on #6265.

## General API

`chainer.mixed16` is a new dtype-like object. It can be passed to any numpy/cupy functions, where the dtype is interpreted as `float16`. Therefore, most existing code will just behave like the conventional FP16 mode. Specifying environment variable `CHAINER_DTYPE=mixed16` is equivalent to `chainer.global_config.dtype = chainer.mixed16`.

`chainer.get_dtype()` is extended to handle `mixed16` in a special manner. By passing an option `max_mixed16=np.float32`, we can get `float32` when the given (or default) dtype is `mixed16`. In this way, developers can choose the appropriate dtype when the users specify `mixed16`. Note that when `float16` is directly specified (either via `dtype` argument or `config.dtype`), `get_dtype()` still returns `float16` regardless of the `map_mixed16` option; i.e., `map_mixed16` is only used in `mixed16` case.

## Mixed precision batch normalization

cuDNN's batch normalization requires float32 beta/gamma parameters for mixed precision computation. Currently, Chainer (strictly speaking, CuPy) automatically casts these parameters to float32 on demand, which may cause a prohibitive overhead. In this PR, `L.BatchNormalization` is fixed to store these parameters in FP32 when `mixed16` dtype is specified, with which the conversion is not needed. Running mean and variance are also stored in float32 on mixed16 mode. `F.batch_normalization` and `F.fixed_batch_normalization` are fixed accordingly to accept arguments with mixed precisions.

### CODE REVIEW

**Comment by gwtnb** [comment]
Jenkins, test this please

**Comment by toslunar** [comment]
The design looks good to me.

**Comment by beam2d** [comment]
Jenkins, test this please.

**Comment by beam2d** [comment]
Jenkins, test this please.

**Comment by toslunar** [comment]
The Jenkins failure is unrelated to the PR.

**Comment by toslunar** [comment]
Jenkins, test this please.

**Comment by toslunar** [comment]
Jenkins, test this please.

**Comment by toslunar** [comment]
Jenkins, test this please.

**Comment by toslunar** [comment]
Jenkins, test this please.

**Comment by toslunar** [comment]
Jenkins, test this please.

**Comment by toslunar** [comment]
Jenkins, test this please.

**Comment by mingxiaoh** [comment]
there is about 20% Convenet Performance regression on resnet50/101/152, googlenetv2/v3 and unet based on CPU-iDeep path, not check GPU yet.  Any comments here? @niboshi
